umbraco/Umbraco-CMS
# V8: Copy on property level

### Prerequisites


If there's an existing issue for this PR then this fixes #5551

### Description

For background info on this PR see #5551 and #5599.

This PR introduces the option to define a property editor as copyable. It's done by setting `CanCopy = true` in the `DataEditor` attribute. This results in a copy icon being displayed next to the property label:

![image](https://user-images.githubusercontent.com/7405322/59963085-62123b00-94ee-11e9-96d0-040838cd97c4.png)

It is of course up to the specific property editor implementation to handle the copy action. The backoffice simply broadcasts a `propertyCopy` event within the editor scope for the property editor controller to handle - something like this:

```js
$scope.$on("propertyCopy", function (ev, args) {
  // implement copy handling here
});
```

This PR also implements copy handling for Nested Content. It works like this:

![nc-copy](https://user-images.githubusercontent.com/7405322/59963137-3e032980-94ef-11e9-80b6-49e632d99399.gif)

#### Limitations

Obviously this feature doesn't work if the property label is hidden. As discussed with @nielslyngsoe on #5599, we need to supply a means to change the property label placement instead of hiding it, so property editors can make use of the full screen width and still have their label. This will be a separate feature/PR. 

#### Future

If this is merged in, the obvious next step would be to implement similar handling for Grid (and possibly other property editors that can be time-consuming to work with).

#### Testing this PR

1. Ensure that all items on a Nested Content property can be copied to another Nested Content property on content of the same type.
2. For language variant content where the Nested Content property is allowed to vary by language, ensure that all items can be copied between languages.
3. For language variant content where the Nested Content property is *not* allowed to vary by language, verify that the property copy icon is *only* present on the main language variant.
4. Verify that you can't copy items from one Nested Content to another if one or more of the copied item types aren't allowed on the target property configuration.
5. Ensure that single items can be copied between Nested Content properties (that this PR didn't break the single item copy feature, in other words).
